### PR TITLE
HPCC-20118 LDAP Resource cache incorrectly thinks entries are stale

### PR DESCRIPTION
**Pull request by RussWhitehead**
The CResPermissionsCache::lookup method incorrectly compares the cache
entry time to the current time to determine if it is stale.  It needs
to take into account the cache timeout. The result of this is that cache
entries are never used, requiring a call to LDAP every time. This PR adjusts
the expiration time by the configured timeout

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@wangkx Please review

**Review by wangkx — APPROVED** [review]
@RussWhitehead looks fine.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge. I targeted 6.4.x, in hopes of getting this in a service release along with https://track.hpccsystems.com/browse/HPCC-20111

File: system/security/shared/caching.cpp [hunk 1]
```diff
@@ -79,14 +79,14 @@ int CResPermissionsCache::lookup( IArrayOf<ISecResource>& resources, bool* pFoun
...
+            const time_t timeExpiry = resParamCacheEntry.first + m_pParentCache->getCacheTimeout();
 
-            if (timestamp < tstamp)//entry was not stale during last cleanup but is stale now
+            if (timeExpiry < tstamp)//entry was not stale during last cleanup but is stale now
                 *pFound++ = false;
             else if(!m_pParentCache->isCacheEnabled() && m_pParentCache->isTransactionalEnabled())//m_pParentCache->getOriginalTimeout() == 0)
             {
                 time_t tctime = getThreadCreateTime();
-                if(tctime <= 0 || timestamp < tctime)
+                if(tctime <= 0 || timeExpiry < tctime)
```
Comment (richardkchapman) [inline system/security/shared/caching.cpp]: Not sure I follow the logic here, but it looks like this branch is used when the cache is not enabled - so I'm not sure why the cache timeout should be a factor in deciding what to do

**Comment by richardkchapman** [comment]
@RussWhitehead One question

File: system/security/shared/caching.cpp (hunk 1 above)
Comment (RussWhitehead) [reply system/security/shared/caching.cpp]: The test for the parent cache (CPermissionsCache) being disabled seems odd, but I don't want to change it in a service release. Opened https://track.hpccsystems.com/browse/HPCC-20121 to investigate
@richardkchapman Please merge

**Comment by richardkchapman** [comment]
I can't say I understand what this code is doing.

**Pull request merged**